hoodie-client/src/main/java/com/uber/hoodie/table/UpsertPartitioner.java

```diff
@@ -0,0 +1,324 @@
+                .getLatestDataFilesBeforeOrOn(partitionPath, latestCommitTime.getTimestamp())
+                .collect(Collectors.toList());
+
+            for (HoodieDataFile file : allFiles) {
+                if (file.getFileSize() < hoodieTable.getConfig().getParquetSmallFileLimit()) {
+                    String filename = file.getFileName();
+                    SmallFile sf = new SmallFile();
+                    sf.location = new HoodieRecordLocation(FSUtils.getCommitTime(filename),
+                        FSUtils.getFileId(filename));
+                    sf.sizeBytes = file.getFileSize();
+                    smallFileLocations.add(sf);
+                }
+            }
+        }
+
+        return smallFileLocations;
+    }
+
+    /**
+     * Obtains the average record size based on records written during last commit.
+     * Used for estimating how many records pack into one file.
+     *
+     * @return
+     */
+    private long averageBytesPerRecord() {
+        long avgSize = 0L;
+        HoodieTimeline commitTimeline =
+            hoodieTable.getMetaClient().getActiveTimeline().getCommitTimeline().filterCompletedInstants();
+        try {
+            if (!commitTimeline.empty()) {
+                HoodieInstant latestCommitTime = commitTimeline.lastInstant().get();
+                HoodieCommitMetadata commitMetadata = HoodieCommitMetadata
+                    .fromBytes(commitTimeline.getInstantDetails(latestCommitTime).get());
+                avgSize = (long) Math.ceil(
+                    (1.0 * commitMetadata.fetchTotalBytesWritten()) / commitMetadata
+                        .fetchTotalRecordsWritten());
+            }
+        } catch (Throwable t) {
+            // make this fail safe.
+            logger.error("Error trying to compute average bytes/record ", t);
+        }
+        return avgSize <= 0L ? hoodieTable.getConfig().getCopyOnWriteRecordSizeEstimate() : avgSize;
+    }
+
+    public BucketInfo getBucketInfo(int bucketNumber) {
+        return bucketInfoMap.get(bucketNumber);
+    }
+
+    public List<InsertBucket> getInsertBuckets(String partitionPath) {
+        return partitionPathToInsertBuckets.get(partitionPath);
+    }
+
+    @Override
+    public int numPartitions() {
+        return totalBuckets;
+    }
+
+    @Override
+    public int getPartition(Object key) {
+        Tuple2<HoodieKey, Option<HoodieRecordLocation>> keyLocation = (Tuple2<HoodieKey, Option<HoodieRecordLocation>>) key;
+        if (keyLocation._2().isDefined()) {
+            HoodieRecordLocation location = keyLocation._2().get();
+            return updateLocationToBucket.get(location.getFileId());
+        } else {
+            final List<InsertBucket> insertBuckets = partitionPathToInsertBuckets.get(keyLocation._1().getPartitionPath());
+            final int insertBucketIndex = Math.floorMod(keyLocation._1().getRecordKey().hashCode(), insertBuckets.size());
```

Comment:
if the weights are gone, how do we ensure the small files are not expanded beyond the maximum configured file size?  This will have effects for a future update.  We need to handle this case.